src/spikeinterface/generation/hybrid_tools.py

```diff
@@ -0,0 +1,518 @@
+            assert len(unit_locations) == num_units, "unit_locations and num_units should have the same length"
+        templates_array = generate_templates(
+            channel_locations,
+            unit_locations,
+            sampling_frequency,
+            ms_before,
+            ms_after,
+            upsample_factor=upsample_factor,
+            seed=seed,
+            dtype=dtype,
+            **generate_templates_kwargs,
+        )
+        nbefore = int(ms_before * sampling_frequency / 1000.0)
+        nafter = int(ms_after * sampling_frequency / 1000.0)
+        templates_ = Templates(templates_array, sampling_frequency, nbefore, True, None, None, None, probe)
+    else:
+        assert isinstance(templates, Templates), "templates should be a Templates object"
+        assert (
+            templates.num_channels == recording.get_num_channels()
+        ), "templates and recording should have the same number of channels"
+        num_units = templates.num_units
+        nbefore = templates.nbefore
+        nafter = templates.nafter
+        unit_locations = compute_monopolar_triangulation(templates)
+
+        channel_locations_rel = channel_locations - channel_locations[0]
+        templates_locations = templates.get_channel_locations()
+        templates_locations_rel = templates_locations - templates_locations[0]
+
+        if not np.allclose(channel_locations_rel, templates_locations_rel):
+            warnings.warn("Channel locations are different between recording and templates. Interpolating templates.")
+            templates_array = np.zeros(templates.templates_array.shape, dtype=dtype)
+            for i in range(len(templates_array)):
+                src_template = templates.templates_array[i][np.newaxis, :, :]
+                templates_array[i] = interpolate_templates(src_template, templates_locations_rel, channel_locations_rel)
+        else:
+            templates_array = templates.templates_array
+
+        # manage scaling of templates
+        templates_ = templates
+        if recording.has_scaleable_traces():
+            if templates_in_uV:
+                templates_array = (templates_array - recording.get_channel_offsets()) / recording.get_channel_gains()
+                # make a copy of the templates and reset templates_array (might have scaled templates)
+                templates_ = templates.select_units(templates.unit_ids)
+                templates_.templates_array = templates_array
+
+    if sorting is None:
+        generate_sorting_kwargs = generate_sorting_kwargs.copy()
+        generate_sorting_kwargs["durations"] = durations
+        generate_sorting_kwargs["sampling_frequency"] = sampling_frequency
+        generate_sorting_kwargs["seed"] = seed
+        generate_sorting_kwargs["num_units"] = num_units
+        sorting = generate_sorting(**generate_sorting_kwargs)
+    else:
+        num_units = sorting.get_num_units()
+        assert sorting.sampling_frequency == sampling_frequency
+
```

Comment:
Are we checking that num_units from the sorting is the same as the number of templates ?

Reply:
done